ceph/mds.go

```diff
@@ -0,0 +1,173 @@
+package ceph
+
+import (
+	"encoding/json"
+	"fmt"
+	"os/exec"
+	"strconv"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	cephCmd                      = "/usr/bin/ceph"
+	mdsBackgroundCollectInterval = 5 * time.Minute
+)
+
+const (
+	MDSModeDisabled   = 0
+	MDSModeForeground = 1
+	MDSModeBackground = 2
+)
+
+type mdsStat struct {
+	FSMap struct {
+		Filesystems []struct {
+			MDSMap struct {
+				FSName string `json:"fs_name"`
+				Info   map[string]struct {
+					GID   uint   `json:"gid"`
+					Name  string `json:"name"`
+					Rank  int    `json:"rank"`
+					State string `json:"state"`
+				} `json:"info"`
+			} `json:"mdsmap"`
+		} `json:"filesystems"`
+	} `json:"fsmap"`
+}
+
+// runMDSStat will run mds stat and get all info from the MDSs within the ceph cluster.
+func runMDSStat(config, user string) ([]byte, error) {
+	var (
+		out []byte
+		err error
+	)
+
+	if out, err = exec.Command(cephCmd, "-c", config, "-n", fmt.Sprintf("client.%s", user), "mds", "stat", "--format", "json").Output(); err != nil {
+		return nil, err
+	}
+
+	return out, nil
+}
+
+// MDSCollector collects metrics from the MDS daemons.
+type MDSCollector struct {
+	config     string
+	user       string
+	background bool
+	logger     *logrus.Logger
+
+	// MDSState reports the state of MDS process running.
+	MDSState *prometheus.Desc
+
+	runMDSStatFn func(string, string) ([]byte, error)
+}
+
+// NewMDSCollector creates an instance of the MDSCollector and instantiates
+// the individual metrics that we can collect from the MDS daemons.
+func NewMDSCollector(exporter *Exporter, background bool) *MDSCollector {
+	labels := make(prometheus.Labels)
+	labels["cluster"] = exporter.Cluster
+
+	mds := &MDSCollector{
+		config:       exporter.Config,
+		user:         exporter.User,
+		background:   background,
+		logger:       exporter.Logger,
+		runMDSStatFn: runMDSStat,
+
+		MDSState: prometheus.NewDesc(
+			fmt.Sprintf("%s_%s", cephNamespace, "mds_daemon_state"),
+			"MDS Daemon State",
+			[]string{"fs", "name", "rank", "state"},
+			labels,
+		),
+	}
+
+	return mds
+}
+
+func (m *MDSCollector) collectorList() []prometheus.Collector {
+	return []prometheus.Collector{}
+}
+
+func (m *MDSCollector) descriptorList() []*prometheus.Desc {
+	return []*prometheus.Desc{
+		m.MDSState,
+	}
+}
+
+func (m *MDSCollector) backgroundCollect(ch chan<- prometheus.Metric) error {
+	for {
+		m.logger.WithField("background", m.background).Debug("collecting MDS stats")
+		err := m.collect(ch)
+		if err != nil {
+			m.logger.WithField("background", m.background).WithError(err).Error("error collecting MDS stats")
+		}
+		time.Sleep(mdsBackgroundCollectInterval)
+	}
+}
+
+func (m *MDSCollector) collect(ch chan<- prometheus.Metric) error {
+	data, err := m.runMDSStatFn(m.config, m.user)
+	if err != nil {
+		return fmt.Errorf("failed getting mds stat: %w", err)
+	}
+
+	ms := &mdsStat{}
+
+	err = json.Unmarshal(data, &ms)
```

Comment:
Should this really unmarshal to pointer to pointer?

Reply:
That's a pretty good catch, although apparently Go dereferences it correctly since my test passes. :) I will fix this.

Reply:
Fixed!